### PR TITLE
Get difference between crates.io hosted code and code in git source

### PR DESCRIPTION
**Pull request by nasifimtiazohi**
1. Infer tag in the git repository for a certain version through string matching heuristics
2. Downloads code from crates.io and compare with the tagged version in git source 
3. Determines crates.io and git source are different if there are new files added or files modified in the crates.io than in the git source [ignoring config files]

### CODE REVIEW

**Review by xvschneider — DISMISSED** [review]
looks good!

File: analysis/src/diff.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            .collect();
+        if entries.len() != 1 {
+            return Err(anyhow!("Error in locating directory for unpacked files"));
+        }
+
+        // Return the directory containing unpacked files
+        Ok(entries[0].path())
+    }
+
+    fn decompress_targz(targz_path: &PathBuf, dest_path: &PathBuf) -> Result<()> {
```
Comment (bmwill) [inline analysis/src/diff.rs]: Similar to the `&String` `&str` thing, we should change these to be `&Path`

File: analysis/src/diff.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        // println!("{:?}", path);
+        assert_eq!(path.exists(), true);
+    }
+
+    #[test]
+    fn test_diff_crate_source_path() {
+        let diff_analyzer = get_test_diff_analyzer();
+        let name = String::from("libc");
+        let version = String::from("0.2.97");
+        let path = diff_analyzer.get_cratesio_version(&name, &version).unwrap();
```
Comment (bmwill) [inline analysis/src/diff.rs:392]: You'll see that change the function signature will stop you needing to make a full `String` here for name and version

File: analysis/src/diff.rs [hunk 3]
```diff
@@ -94,13 +101,58 @@ impl DiffAnalyzer {
...
+        // as a follow-up heuristic, look for the presence of crate name in the tag
+        for tag in candidate_tags.iter() {
+            let tag = tag.ok_or_else(|| anyhow!("Error in fetching tags"))?;
+            if tag.contains(name) {
+                let oid = hm
+                    .get(tag)
+                    .ok_or_else(|| anyhow!("Error in fetching tags"))?;
+                return Ok(Some(oid.clone()));
+            }
+        }
```
Comment (bmwill) [inline analysis/src/diff.rs]: Wow nice job thinking through these corner cases. I see you have a single test below it might be nice have a few more tests to verify you've handled these edge cases properly.

File: analysis/src/diff.rs [hunk 4]
```diff
@@ -178,16 +188,58 @@ impl DiffAnalyzer {
...
+        let mut format = git2::DiffStatsFormat::NONE;
+        format |= git2::DiffStatsFormat::FULL;
+        let buf = stats.to_buf(format, 80)?;
+        print!(
+            "difference between crates.io and source is:\n {}",
+            std::str::from_utf8(&*buf).unwrap()
+        );
+
+        Ok(diff)
+    }
```
Comment (bmwill) [inline analysis/src/diff.rs]: One potential issue with this way of diffing crates is that it probably doesn't work 100% correctly with crates that are a part of a workspace. Eg you should take a look at running this on the guppy crate since that repo has a handful of crates in a workspace.

File: analysis/src/diff.rs [hunk 5]
```diff
@@ -241,4 +293,29 @@ mod test {
         let oid = diff_analyzer.get_head_commit_oid_for_version(&repo, &name, &version);
         assert_eq!(oid.is_ok(), true);
     }
+
+    #[test]
+    fn get_test_analyze_diff() {
+        let diff_analyzer = get_test_diff_analyzer();
+        let name = String::from("diffy");
+        let url = "https://github.com/bmwill/diffy";
```
Comment (bmwill) [inline analysis/src/diff.rs:427]: <3

File: analysis/src/diff.rs [hunk 6]
```diff
@@ -35,39 +53,102 @@ impl DiffAnalyzer {
...
-        Ok(Some(DiffReport {
-            ..Default::default()
-        }))
+        // Get release version commit within crate repo
+        let git_version_commit = match crate_repo.find_commit(head_commit_oid) {
+            Ok(commit) => commit,
+            Err(_error) => {
+                // It is possilbe that the commit object does not belong to any branch in the repository,
+                // e.g., syn 0.15.44
+                // TODO: can we get around this to still get the diff?
```
Comment (bmwill) [inline analysis/src/diff.rs:103]: Would you be able to provide a bit more context on this error you're seeing?

File: analysis/src/diff.rs (hunk 4 above)
Comment (nasifimtiazohi) [reply analysis/src/diff.rs]: I thought about those cases. My thinking was that in those cases the diff would be files that are in git source and not in crates.io such as the files in other crates of the workspace. Now for crate_source_diff analysis, we are already ignoring all files that are in git source and not in crates.io as that could arise for various reasons like this. 

However, if there are files in crates.io but not in git source and files in both places but different, those are the red flags. On that heuristic, this analysis is simple yet effective. That's my rationale here. Let me know if you have any suggestion.

File: analysis/src/diff.rs (hunk 6 above)
Comment (nasifimtiazohi) [reply analysis/src/diff.rs:103]: https://github.com/dtolnay/syn/commit/6d798b63c255e90b7b1dbbfb3707fdce1704a18d - I checked on the local git as well that the commit does not belong to any branch. My guess is that the commit/branch was later squashed or deleted. Therefore, `repo.find_commit()` fails to find to commit given the sha.

File: analysis/src/diff.rs (hunk 4 above)
Comment (bmwill) [reply analysis/src/diff.rs]: So the issue I'm worried about is the differing directory level, for instance:

the `guppy` crate lives in the repo `https://github.com/facebookincubator/cargo-guppy` but it doesn't live at the root of the repository, its located at `/guppy` where as when you're looking at the code from the crate you've downloaded from crates.io, that code will be located at the root of the project `/`. So in that instance wouldn't all the files be different?

I'm really just theorizing here and perhaps `git2`'s diff functionality is smart enough to do the rename detection when diffing. Would you be able to add a test for this particular instance so that we can be confident your solution covers this case?

Comment (nasifimtiazohi) [reply analysis/src/diff.rs]: Oh that's a good catch. Yes, I didn't think about the path difference issue. Let me check.

File: analysis/src/diff.rs (hunk 6 above)
Comment (bmwill) [reply analysis/src/diff.rs:103]: Interesting...One thing that we could try, before bailing out, would be to take the tag name and try to fetch the tag directly. Doing the equivalent of `git fetch <SHA | tag name>` since presumably you have the tag name here or have been able to determine the commit SHA.

File: analysis/src/diff.rs [hunk 7]
```diff
@@ -225,4 +297,53 @@ mod test {
         assert_eq!(repo.path().exists(), true);
         // TODO add tests for non-git repos
     }
+
+    #[test]
+    fn test_diff_head_commit_oid_for_version() {
```
Comment (bmwill) [inline analysis/src/diff.rs:522]: Awesome tests

File: analysis/src/diff.rs [hunk 8]
```diff
@@ -232,6 +234,28 @@ impl DiffAnalyzer {
...
+
+    fn init_git(&self, path: &Path) -> Result<Repository> {
+        // initiates a git repository in the path
+        let repo = Repository::init(path)?;
+
+        // add and commit existing files
+        let mut index = repo.index()?;
+        index.add_all(["*"].iter(), IndexAddOption::DEFAULT, None)?;
+        let oid = index.write_tree()?;
+        let signature = Signature::now("Nasif Imtiaz", "nasifimtiaz@fb.com")?;
```
Comment (bmwill) [inline analysis/src/diff.rs:305]: You could probably just have this be a random place holder instead of your name and email

File: analysis/src/diff.rs [hunk 9]
```diff
@@ -281,6 +316,43 @@ impl DiffAnalyzer {
...
+        // Given a crate name and its repository
+        // This function returns the path to Cargo.toml for the given crate
+
+        let path = repo
+            .path()
+            .parent()
+            .ok_or_else(|| anyhow!("Fatal: .git file has no parent"))?;
+
+        // TODO: Why would guppy fail to build a graph?
+        // e,g, for combine 3.8.1, guppy fails
```
Comment (bmwill) [inline analysis/src/diff.rs:352]: What types of errors did you see?

**Review by bmwill — DISMISSED** [review]

**Review by bmwill — APPROVED** [review]
